### ISSUE TITLE
ValueError while parsing data  v1.4.2 

### ISSUE THREAD

**Issue by leighton** [pedrokiefer/aiostomp#18]
I've come across this. It indicates that `b'\x00'` is not present in string and so a single value is returned.
```
protocol: <aiostomp.aiostomp.StompReader object at 0x7f8b80c6ea58>
transport: <_SelectorSocketTransport fd=6 read=polling write=<idle, bufsize=0>>
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/asyncio/selector_events.py", line 824, in _read_ready__data_received
    self._protocol.data_received(data)
  File "/usr/local/lib/python3.7/site-packages/aiostomp/aiostomp.py", line 393, in data_received
    self._protocol.feed_data(data)
  File "/usr/local/lib/python3.7/site-packages/aiostomp/protocol.py", line 40, in feed_data
    pending_data = self._feed_data(data)
  File "/usr/local/lib/python3.7/site-packages/aiostomp/protocol.py", line 57, in _feed_data
    b''.join(self._pending_parts))
  File "/usr/local/lib/python3.7/site-packages/aiostomp/protocol.py", line 133, in _parse_data
    _, extra = data[missing_length:].split(self.EOF, 1)
ValueError: not enough values to unpack (expected 2, got 1)
```

**Comment by pedrokiefer** [comment]
Hi @leighton, a few questing regarding this bug:
Could you please send what is on `self._pending_parts` ? 
What message broker are you using? 
Is the `content-length` header with right length of the data you are sending?

**Comment by HerrBerg** [comment]
Hi Pedro,
after some more high volume testing I also ran into this issue:
here is some info when catching the exception:

missing_length: `957`
data: ```b'AgICAgICAgICAgICAgICAg\r\nICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAg\r\nICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAg\r\nICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgCjw/eHBh\r\nY2tldCBlbmQ9InciPz4KZW5kc3RyZWFtCmVuZG9iagp4cmVmCjAgMjAKMDAwMDAwMDAwMCA2NTUz\r\nNSBmIAowMDAwMDAwMDE1IDAwMDAwIG4gCjAwMDAwMDAxMDUgMDAwMDAgbiAKMDAwMDAwMDI5MyAw\r\nMDAwMCBuIAowMDAwMDAwMzY2IDAwMDAwIG4gCjAwMDAwMDA1OTEgMDAwMDAgbiAKMDAwMDAwMTAz\r\nNCAwMDAwMCBuIAowMDAwMDAxNTY0IDAwMDAwIG4gCjAwMDAwMDE3NDUgMDAwMDAgbiAKMDAwMDAw\r\nMTkwMCAwMDAwMCBuIAowMDAwMDAyMjc2IDAwMDAwIG4gCjAwMDAwMDI0NTIgMDAwMDAgbiAKMDAw\r\nMDAwMjYxNSAwMDAwMCBuIAowMDAwMDAyNzgyIDAwMDAwIG4gCjAwMDAwMDI4MzYgMDAwMDAgbiAK\r\nMDAwMDAwMzAyMyAwMDAwMCBuIAowMDAwMDA0NDY3IDAwMDAwIG4gCjAwMDAwMDUzNzEgMDAwMDAg\r\nbiAKMDAwMDAwNTg5MSAwMDAwMCBuIAowMDAwMDA1OTUxIDAwMDAwIG4gCnRyYWlsZXIKPDwKL1Np\r\nemUgMjAKL1Jvb3QgMSAwIFIKL0luZm8gMiAwIFIKPj4Kc3RhcnR4cmVmCjgwNjkKJSVFT0YK\r\n\x00\nMESSAGE\ncontent-length:4000\nexpires:0\ndestination:/topic/streams.email.json-black-dedup\nsubscription:1\npriority:4\nActiveMQ.MQTT.QoS:1\nmessage-id:ID\\csmtp01-35563-1549820520348-3\\c692\\c-1\\c1\\c3614\npersistent:true\ntimestamp:1551188141852\n\n{"smtp_mail_from":"fyri399@xxxxxx.xxxxxxx","data_colorcode":"black","email_attachment_count":"0","source_ip":"xxx.xxx.xxx.xxx","detected_text_language":"de","email_subject":"Guten Tag \\uD83D\\uDCA1","email_attachment_content_types":[],"data_origin":"com.xxxxxx.spam.xxxxxxxxx","email_urls":["https://abpodoja1987.blogspot.md/"],"smtp_timestamp":"Tue, 26 Feb 2019 13:35:39 +0000","email_headers_raw":{"date":["Tue, 26 Feb 2019 06:37:29 -0700"],"mime-version":["1.0"],"subject":["=?UTF-8?Q?Guten_Tag_=F0=9F=92=A1?="],"received":["from mail.xxxxxx.xxxxxxx (16.185.xxxxxxxxxxxx [xxx.xxx.xxx.xxx])\\r\\n\\tby example.me (Haraka/2.8.23) with ESMTPA id 628BD19C-69D3-4936-9948-CEC66A7D8362.1\\r\\n\\tenvelope-from <fyri399@xxxxxx.xxxxxxx> (authenticated bits=0);\\r\\n\\tTue, 26 Feb 2019 13:35:37 +0000"],"content-transfer-encoding":["7bit"],"received-spf":["None (example.me: domain of xxxxxx.xxxxxxx does not designate xxx.xxx.xxx.xxx as permitted sender) receiver=example.me; identity=mailfrom; client-ip=xxx.xxx.xxx.xxx; helo=mail.xxxxxx.xxxxxxx; envelope-from=<fyri399@xxxxxx.xxxxxxx>","None (example.me: domain of mail.xxxxxx.xxxxxxx does not designate xxx.xxx.xxx.xxx as permitted sender) receiver=example.me; identity=helo; client-ip=xxx.xxx.xxx.xxx; helo=mail.xxxxxx.xxxxxxx; envelope-from=<fyri399@xxxxxx.xxxxxxx>"],"message-id":["<3bd8afdb-b03e-4729-b45e-d8a5cda8eb46@xxxxxx.xxxxxxx>"],"from":["die_blumentopferde <fyri399@xxxxxx.xxxxxxx>"],"content-type":["text/plain; charset=utf-8; format=flowed"],'```
len(data): `2690`
self.pending_parts: ```[b'AgICAgICAgICAgICAgICAg\r\nICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAg\r\nICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAg\r\nICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgCjw/eHBh\r\nY2tldCBlbmQ9InciPz4KZW5kc3RyZWFtCmVuZG9iagp4cmVmCjAgMjAKMDAwMDAwMDAwMCA2NTUz\r\nNSBmIAowMDAwMDAwMDE1IDAwMDAwIG4gCjAwMDAwMDAxMDUgMDAwMDAgbiAKMDAwMDAwMDI5MyAw\r\nMDAwMCBuIAowMDAwMDAwMzY2IDAwMDAwIG4gCjAwMDAwMDA1OTEgMDAwMDAgbiAKMDAwMDAwMTAz\r\nNCAwMDAwMCBuIAowMDAwMDAxNTY0IDAwMDAwIG4gCjAwMDAwMDE3NDUgMDAwMDAgbiAKMDAwMDAw\r\nMTkwMCAwMDAwMCBuIAowMDAwMDAyMjc2IDAwMDAwIG4gCjAwMDAwMDI0NTIgMDAwMDAgbiAKMDAw\r\nMDAwMjYxNSAwMDAwMCBuIAowMDAwMDAyNzgyIDAwMDAwIG4gCjAwMDAwMDI4MzYgMDAwMDAgbiAK\r\nMDAwMDAwMzAyMyAwMDAwMCBuIAowMDAwMDA0NDY3IDAwMDAwIG4gCjAwMDAwMDUzNzEgMDAwMDAg\r\nbiAKMDAwMDAwNTg5MSAwMDAwMCBuIAowMDAwMDA1OTUxIDAwMDAwIG4gCnRyYWlsZXIKPDwKL1Np\r\nemUgMjAKL1Jvb3QgMSAwIFIKL0luZm8gMiAwIFIKPj4Kc3RhcnR4cmVmCjgwNjkKJSVFT0YK\r\n\x00\nMESSAGE\ncontent-length:4000\nexpires:0\ndestination:/topic/streams.email.json-black-dedup\nsubscription:1\npriority:4\nActiveMQ.MQTT.QoS:1\nmessage-id:ID\\csmtp01-35563-1549820520348-3\\c692\\c-1\\c1\\c3614\npersistent:true\ntimestamp:1551188141852\n\n{"smtp_mail_from":"fyri399@xxxxxx.xxxxxxx","data_colorcode":"black","email_attachment_count":"0","source_ip":"xxx.xxx.xxx.xxx","detected_text_language":"de","email_subject":"Guten Tag \\uD83D\\uDCA1","email_attachment_content_types":[],"data_origin":"com.xxxxxx.spam.xxxxxxxxx","email_urls":["https://abpodoja1987.blogspot.md/"],"smtp_timestamp":"Tue, 26 Feb 2019 13:35:39 +0000","email_headers_raw":{"date":["Tue, 26 Feb 2019 06:37:29 -0700"],"mime-version":["1.0"],"subject":["=?UTF-8?Q?Guten_Tag_=F0=9F=92=A1?="],"received":["from mail.xxxxxx.xxxxxxx (16.185.xxxxxxxxxxxx [xxx.xxx.xxx.xxx])\\r\\n\\tby example.me (Haraka/2.8.23) with ESMTPA id 628BD19C-69D3-4936-9948-CEC66A7D8362.1\\r\\n\\tenvelope-from <fyri399@xxxxxx.xxxxxxx> (authenticated bits=0);\\r\\n\\tTue, 26 Feb 2019 13:35:37 +0000"],"content-transfer-encoding":["7bit"],"received-spf":["None (example.me: domain of xxxxxx.xxxxxxx does not designate xxx.xxx.xxx.xxx as permitted sender) receiver=example.me; identity=mailfrom; client-ip=xxx.xxx.xxx.xxx; helo=mail.xxxxxx.xxxxxxx; envelope-from=<fyri399@xxxxxx.xxxxxxx>","None (example.me: domain of mail.xxxxxx.xxxxxxx does not designate xxx.xxx.xxx.xxx as permitted sender) receiver=example.me; identity=helo; client-ip=xxx.xxx.xxx.xxx; helo=mail.xxxxxx.xxxxxxx; envelope-from=<fyri399@xxxxxx.xxxxxxx>"],"message-id":["<3bd8afdb-b03e-4729-b45e-d8a5cda8eb46@xxxxxx.xxxxxxx>"],"from":["die_blumentopferde <fyri399@xxxxxx.xxxxxxx>"],"content-type":["text/plain; charset=utf-8; format=flowed"],']```
len(self.pending_parts): `1`
len(self.pending_parts[0]): `2690`
content_length: `11696`

so `data[:missing_length]` ends with `self.EOF` and in `data[missing_length:]` there is no `self.EOF` anymore.

**Comment by pedrokiefer** [comment]
Just checked the spec, and we are doing something wrong here :(

Content-length header has size of all the frame body AND the ending null char. At least the error message on the spec does this.

```
x = b"""ERROR
receipt-id:message-12345
content-type:text/plain
content-length:171
message: malformed frame received

The message:
-----
MESSAGE
destined:/queue/a
receipt:message-12345

Hello queue a!
-----
Did not contain a destination header, which is REQUIRED
for message propagation.
\x00"""
headers, body = x.split(b'\n\n', 1)
len(body)
171
```

**Comment by pedrokiefer** [comment]
@leighton @HerrBerg can you please check if #19 fixes this issues for you?

**Comment by HerrBerg** [comment]
It's now crashing with `data` always starting with b'\x00\nMESSAGE\ncontent-length:....' and `missing_length` being zero

**Comment by HerrBerg** [comment]
I now had a case where missing_length was != 0:
`missing_length = 10536`
and `data[missing_length - 2:]` (minus *two*) starting with `b'\x00\nMESSAGE\ncontent-length:`

**Comment by pedrokiefer** [comment]
I should definitely rewrite this part of the code. There are way too many conditions and edge cases.

I've pushed another try on fixing this issue. If that doesn't work, i'll rewrite this parsing chaos.

**Comment by pedrokiefer** [comment]
@HerrBerg I've rewrote the parser, based it on the ActiveMQ parser, all tests are passing, still have to test it with some server (ActiveMQ and RabbitMQ). The code is on #20.

**Comment by leighton** [comment]
@pedrokiefer I've been testing the code on #20 on a high volume feed for the last 2 days (~1M messages). The error no longer occurs and another frequent error, associated with not being able to decode a part of message, has also been resolved.

**Comment by pedrokiefer** [comment]
I’ll merge #20 and create a new release!

Thanks for testing it!

**Comment by HerrBerg** [comment]
I also haven't seen any issues while receiving ~4M messages.

**Comment by pedrokiefer** [comment]
Just released version 1.5.0 with the new parser.
